File: lab3/fibo.c
#include <stdio.h>
#include <conio.h>
long int table[100] = {0};

long int fibo(int n)
{
    if (n == 1 || n == 2)
        return 1;
    else
    {
        if (table[n] == 0)
        {
            table[n] = fibo(n - 1) + fibo(n - 2);
        }
        return table[n];
    }
}
int main(){
    int num;
    printf("Enter the nth number: \n");
    scanf("%d",&num);
    long int result=fibo(num);
    printf("The result of %dth fibonacci number is %ld",num,result);
    return 0;
}

File: lab3assignments/fibo.py
def fibo(n,table=None): # table={} acts as an mutable argument or overwrriten by a new dictionary calls and takes more space ! 
    if table is None:
        table={} 
        
    if n==1 or n==2:
        return 1
    if n in table:
        return table[n]
    table[n]=fibo(n-1,table)+fibo(n-2,table)
    return table[n]
        

n=int(input("Enter the nth term to calculate the fibonacci series: "))
result =fibo(n)
print(f"The nth fibonacci term is {result}")

#The defualt max recurrsion availale is 1000 calls ! 

#The memoization works like this while using a dictionary , if you ever wonder how the above code works you can run this below code hehe

# table={3:'hi'}
# print(table[3])
# table[4]="mi "  
# print(table)
# print(table[4])


